### ISSUE TITLE
Document how to add SSL Certificate

### ISSUE THREAD

**Issue by jbeorse** [odk-x/tool-suite-X#147]
Add documentation for adding an SSL certificate to a Sync Endpoint. Probably should include links to other docs for things like generating them. 

From the 2019 convening: https://docs.google.com/document/d/1_4x9x3eHxo4qevBRlpmCC9L2bVlUkAQuAlidT8wJDzI/edit#heading=h.264josm5lg7

**Comment by wbrunette** [comment]
Process improved to add certbot to sync-endpoint

**Comment by wbrunette** [comment]
Docs update: https://github.com/odk-x/docs/pull/10

**Comment by elmps2018** [comment]
This appears to be at least in part underway but looks like it needs further work and probably is worth pursuing.

**Comment by khushishikhu** [comment]
@elmps2018  Can I try to work on this issue. This will help me to learn more about the documentation of odk-x as a contributor and will research to make this issue done.

**Comment by elmps2018** [comment]
Hi @khushishikhu, it looks like you already have a number of open issues. We ask folks to focus on working on a few issues -- if you can finish off some of the other ones then can ask again for this one!

**Comment by khushishikhu** [comment]
@elmps2018 yes sure I understand, that is why I asked it now because recently my PR got merged of one issue and one more is about to done. So I thought I am allowed to take a one now and start working. But It's okay I totally understand your point too.
Thank you!!

**Comment by elmps2018** [comment]
@khushishikhu, looking at assignees quick it looked like you still have more than one open issue assigned. Are there some you are no longer working on?
<img width="1275" alt="Screen Shot 2021-04-15 at 12 47 04 PM" src="https://user-images.githubusercontent.com/35489974/114915069-b8015280-9de8-11eb-9d6c-e8bb99c99b1e.png">

**Comment by wbrunette** [comment]
@khushishikhu our general limit is normally two open with more issues allowed with completed PRs in waiting for reviewer. This means @elmps2018 answered correctly, 

This means the following PRs should have some movement before getting a third, we allowed a higher number sometimes because of interrelated like (#225 and #110). 

However currently, the mentors have answered on:
#225 (and thus related to #110)
#250

This means you have two to work on.

I see you are waiting on @elmps2018 #229

**Comment by khushishikhu** [comment]
@elmps2018 @wbrunette okay sure I got your point. Thank you for making me consern I will take care of this thing from next time.

**Comment by khushishikhu** [comment]
@elmps2018 can I take this issue?

**Comment by elmps2018** [comment]
Hi @khushishikhu! Appreciate your enthusiasm for working on issues! It looks like you still have two open issues 
<img width="1331" alt="Screen Shot 2021-04-17 at 2 46 05 PM" src="https://user-images.githubusercontent.com/35489974/115125123-c2a81d00-9f8b-11eb-9b67-45bfe7236dbb.png">

Are you awaiting review on any of those?

**Comment by wbrunette** [comment]
@elmps2018 (those two are combined) so it's fine for her to have a third.

**Comment by elmps2018** [comment]
@wbrunette @khushishikhu Great thanks for clarifying!

**Comment by wbrunette** [comment]
@khushishikhu since this issue was created we have revised the default setup to automatically obtain a encryption certificate using letsencrypt. However, sometimes people still have their own certificate already and they simply want to add an instance sync-endpoint to an existing domain (e.g., odkx.mydomain.com and mydomain already has a registered TLS/SSL certifcate). Alternatively some people do not want to use letsencrypt. Therefore how to have sync-endpoint reference the correct TLS certifacte is important.

**Comment by khushishikhu** [comment]
@wbrunette I was reading this document https://docs.odk-x.org/sync-endpoint-manual-setup/ and then this https://gist.github.com/chrismclarke/32a0b7331c6027a3e7be4d0d2c0866d3  and found out that there isn't the documentation for creating SSL certificate in [this](https://docs.odk-x.org/sync-endpoint-manual-setup/) .does the instructions given [here](https://gist.github.com/chrismclarke/32a0b7331c6027a3e7be4d0d2c0866d3) are using letsencrypt? and I have to write the documentation for adding own SSL/TLS certificate without using letsencrypt?

**Comment by wbrunette** [comment]
The sync-endpoint-default-setup automatically uses letsencryt and does this for the user in the setup script. 
Therefore, we have the use case of starting from nothing covered/documented, what we are looking for is if you already have an TLS/SSL certificate how do you setup sync-endpoint with a previously obtained certificate. Does that make sense? It's good that you did some investigating as it helps to understand the system.

**Comment by khushishikhu** [comment]
@wbrunette  what does here domain mean, like do I have to create my own one or we have some specified domains that are to be written here? 
![Capture1](https://user-images.githubusercontent.com/65439761/115377899-ac999700-a1ed-11eb-93a3-523c802dbad9.PNG)
 
and I was trying to create a certificate for testing purpose so that I can know about how things work. but then I run this command `sudo certbot certificate` in which if I have to add SSL/TLS certificate it will be saved into letsencrypt file. so if anyone add the SSL certificate to an existing domain will get save to this letsencrypt file? and is this what has to be documented because I think certificates in letsencrypt file doesn't mean that lstsencrypt is automatically generating the certificate.

![Capture](https://user-images.githubusercontent.com/65439761/115377845-a0153e80-a1ed-11eb-9b39-7a1ad7c1c652.PNG)

**Comment by wbrunette** [comment]
You have to own the domain for certbot to work. The domain traffic must be directed to the machine that is running certbot must be on that domain. Certbot uses it's remote servers to verify that you actually own the domain then issues the certificate.

Let's encrypt is a provider of TLS/SSL certificates. The nice thing is that the certificates are free; https://letsencrypt.org/
There are lots of companies that you can obtain a TLS/SSL certificate from, but you have to pay money. Examples include Microsoft, Amazon, GoDaddy, DigiCert, etc. 

I produces a PEM file. Perhaps @matt9j has some good references for you to look at. 

Basically you need to document where to put the certificate someone already obtained.

**Comment by khushishikhu** [comment]
Thank you for making me clear @wbrunette and I think this https://phoenixnap.com/kb/install-ssl-certificate-nginx can be the way to add an owned  SSL certificate in a nginx server ( HTTPS port:443 ) and an existing domain. please let me know your opinion?

**Comment by matt9j** [comment]
Hi @khushishikhu ! That could be a possible approach, but the main challenge I think will be integrating that with how the nginx config and certificates are handled in docker right now. In addition to importing the certificate, the certbot functionality will need to be disabled.

**Comment by wbrunette** [comment]
@matt9j good point. I think the main issue is to document how to get the certificate working with nginx. I think secondary is to open an issue to add to disable certbot. I just created https://github.com/odk-x/tool-suite-X/issues/264. Feel free to comment here.

@khushishikhu write the instructions assuming the certbot can be disabled.

**Comment by khushishikhu** [comment]
@wbrunette @matt9j  Thank you for clearing out my doubts, It really helped me a lot.
